sophus/ceres_local_parameterization.hpp

```diff
@@ -0,0 +1,96 @@
+#ifndef SOPHUS_CERES_LOCAL_PARAMETERIZATION_HPP
+#define SOPHUS_CERES_LOCAL_PARAMETERIZATION_HPP
+
+namespace Sophus {
+
+/// Type trait used to distinguish mappable vector types from scalars
+/// We use this class to distinguish Sophus::Vector<Scalar, N> from Scalar types
+/// in LieGroup<T>::Tangent
+///
+/// Fortunately, ceres::Jet is not mappable
+template <typename T>
+struct is_mappable_type_t {
```

Comment:
I'm not sure I understand the purpose of this trait. Isn't it the same as the negation of `std::is_scalar`?

Reply:
We want to distinguish between two cases:
* `LieGroup::Tangent` is a `Vector<...>`, thus getting `Tangent`-like reference object from pointer to data is constructing `Eigen::Map<...>`
* `LieGroup:Tangent` is a "scalar", thus getting `Tangent`-like reference object from pointer to data is pointer dereferencing



This is not negation of `std::is_scalar` because scalar types in c++ (in other words, types such that `std::is_scalar_v` is true) should be from one of primary type categories: arithmetic, enumeration, pointer, pointer-to-member or `std::nullptr_t`.
But in our case "scalar" sometimes needs to be `ceres::Jet`; and the primary type category corresponding to `ceres::Jet` is non-union class in terms of c++ standard (thus, `std::is_class<ceres::Jet<...>>` is true, and is the only primary type category predicate that is true for `ceres::Jet`).

Only one primary category can correspond to a type, thus `ceres::Jet` cannot be a scalar type in terms of c++.


That's why `std::is_scalar_v<ceres::Jet<...>>` is `false`, which does not fit to our purposes.

Reply:
Thanks for the explanation. This means that the primary logic for distinguishing between a "mappable" and non-"mappable" type is the availability of `Eigen::internal::traits` specialization, correct? I'm asking because your trait is slighlty convoluted can be rewritten in a compact way to be more readable:

```cpp
template<class T, std::size_t = sizeof(T)>
constexpr std::true_type complete(T*);
constexpr std::false_type complete(...);

template<class T>
using IsSpecialized = decltype(complete(std::declval<T*>()));

template<class T>
using HasEigenNumTraits = IsSpecialized<Eigen::NumTraits<std::decay_t<T>>>;
```
As an additional note, `Eigen::internal` is not intended for general use. You could reference `Eigen::NumTraits` instead, no?

Reply:
`Eigen::NumTraits<Eigen::Matrix<...>>` is complete.
In fact, for every type `T` `Eigen:NumTraits<T>` is complete due to https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/NumTraits.h#L234


Having `Eigen::internal::traits` specialized for `T` is prerequisite to be able to construct `Eigen::Map<T>` (that's the reason, for example, why we specialize it for each Lie group, from my understanding).

I've updated code with your approach adapted to `Eigen::internal::traits`

Reply:
But then you should be able to check whether the type is specialized for `Eigen::Map` instead of referencing `Eigen::internal::traits` directly:
```cpp
using IsMappable = IsSpecialized<Eigen::Map<std::decay_t<T>>>;
```

Reply:
```c++
using IsMappable = IsSpecialized<Eigen::Map<std::decay_t<T>>>;
```
This particular check doesn't work, because it requires `Eigen::internal::traits<T>` to be complete _as a prerequisite_.

Due to `Eigen::Map<T>` being derived from `Eigen::MapBase<Map<T,..>>` and `Eigen::MapBase<D,...>` declaring a bunch of compile-time constants via `Eigen::internal::traits<D>`.
https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/Map.h#L96-97
https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/MapBase.h#L39-50

Reply:
You should probably add the reasoning as a comment because it is entirely possible that this will break in the future.

Reply:
I've changed comments/naming around that trait (and added emphasis about the only reason for using `Mapper` structure in tests being having a single code for all groups).